### PR TITLE
Add email controller flow and user entity

### PR DESCRIPTION
**Pull request by boykas**

### CODE REVIEW

File: server-app/src/main/java/ua/com/findcoach/controllers/EmailController.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import java.io.IOException;
+import java.net.URLDecoder;
+
+@Controller
+public class EmailController {
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/controllers/EmailController.java]: It is better to name it corresponding to it's functionality. It is functionality is to authenticate user. Better name is AuthenticationController

File: server-app/pom.xml [hunk 2]
```diff
@@ -9,14 +9,89 @@
...
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <version>1.1.3</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-core</artifactId>
+            <version>4.2.1.RELEASE</version>
```
Comment (senichas) [inline server-app/pom.xml]: Is version in several dependencies is the same like in spring it is better to move version in property and use property value.

File: server-app/pom.xml [hunk 3]
```diff
@@ -9,14 +9,89 @@
...
+            <artifactId>hibernate-annotations</artifactId>
+            <version>3.5.6-Final</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>9.4-1203-jdbc42</version>
+        </dependency>
+
+        <!-- Tests -->
```
Comment (senichas) [inline server-app/pom.xml]: If library is used for testing it should have scope = test

File: server-app/src/main/java/ua/com/findcoach/api/EmailPojo.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package ua.com.findcoach.api;
+
+/**
+ * Created by DENIS on 19.09.2015.
+ */
+public class EmailPojo {
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/api/EmailPojo.java]: Class should be renamed AuthenticationResponse

File: server-app/src/main/java/ua/com/findcoach/controllers/EmailController.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @RequestMapping(method = RequestMethod.POST, value = "email")
+    @ResponseBody
+    public String postAnswer(@RequestBody String body) throws JsonMappingException, JsonParseException, IOException {
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/controllers/EmailController.java]: It is better to name method authenticate. And if request JSON has the same structure as AuthenticationRequest you can write like 
authenticate(@RequestBody AuthentictionRequest requestData) 
Actually I am not sure about it but seems I've done it in this way. If it is not work let me know I will try.

File: server-app/src/main/java/ua/com/findcoach/domain/User.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    public boolean isActive() {
+        return isActive;
+    }
+
+    public void setActive(boolean isActive) {
+        this.isActive = isActive;
+    }
+
+    public static void main(String[] args){
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/domain/User.java]: What is it?

File: server-app/src/main/webapp/WEB-INF/applicationContext.xml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (senichas) [inline server-app/src/main/webapp/WEB-INF/applicationContext.xml]: Why we need two context files?

File: server-app/src/main/webapp/WEB-INF/web.xml [hunk 8]
```diff
@@ -1,18 +1,18 @@
...
+         id="WebApp_ID" version="3.0">
     <display-name>All Kiev</display-name>
 
-    <context-param>
-        <param-name>contextConfigLocation</param-name>
-        <param-value>
-            /WEB-INF/root-applicationContext.xml
-        </param-value>
-    </context-param>
+    <!--<context-param>-->
```
Comment (senichas) [inline server-app/src/main/webapp/WEB-INF/web.xml]: Comments should be removed.

File: server-app/src/tests/EmailControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+/**
+ * Created by DENIS on 20.09.2015.
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@WebAppConfiguration
+@ContextConfiguration(locations = { "file:src/main/webapp/WEB-INF/defaultServlet-context.xml"})
+public class EmailControllerTest {
```
Comment (senichas) [inline server-app/src/tests/EmailControllerTest.java]: Let us skip unit test for controller on this step of developent.

File: server-app/src/main/java/ua/com/findcoach/domain/User.java (hunk 6 above)
Comment (boykas) [inline server-app/src/main/java/ua/com/findcoach/domain/User.java]: getters and setters

File: server-app/src/main/webapp/WEB-INF/applicationContext.xml (hunk 7 above)
Comment (boykas) [inline server-app/src/main/webapp/WEB-INF/applicationContext.xml]: i didn't know that spring uses applicationContext.xml by default.

File: server-app/src/main/java/dao/UserDaoInterface.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package dao;
+
+import ua.com.findcoach.domain.User;
+
+/**
+ * Created by DENIS on 21.09.2015.
+ */
+public interface UserDaoInterface {
```
Comment (senichas) [inline server-app/src/main/java/dao/UserDaoInterface.java]: We don't use type (interface) in type name. The best way to name this class is UserDao

File: server-app/src/main/java/ua/com/findcoach/controllers/AuthenticationController.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.io.IOException;
+import java.net.URLDecoder;
+
+@Controller
+public class AuthenticationController {
+    @RequestMapping(method = RequestMethod.POST, value = "email")
+    @ResponseBody
+    public String postAnswer(@RequestBody String body) throws JsonMappingException, JsonParseException, IOException {
+        EmailValidator emailValidator = new EmailValidator();
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/controllers/AuthenticationController.java]: We use spring for creation and wiring objects. Put EmailValidator in spring context and make it autowire here.

File: server-app/src/main/java/ua/com/findcoach/controllers/AuthenticationController.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        ObjectMapper mapper = new ObjectMapper();
+        String decodeJSON = new URLDecoder().decode(body, "UTF-8");
+        System.out.println(decodeJSON);
+        AuthenticationRequest pojo = mapper.readValue(decodeJSON, AuthenticationRequest.class);
+        System.out.println(pojo.getEmail());
+        if (!emailValidator.validate(pojo.getEmail())) {
+            return "You wrote wrong email";
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/controllers/AuthenticationController.java]: We have to reply with AuthenticateResponse. Response should contains fields boolean result and string message. If okay then result true and message empty if not result false and message contains message to display on UI

File: server-app/src/main/java/ua/com/findcoach/controllers/EmailValidator.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+public class EmailValidator {
+    private Pattern pattern;
+    private Matcher matcher;
+
+    private static final String EMAIL_PATTERN =
+            "^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@" +
+                    "[A-Za-z0-9-]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$";
+
+    public EmailValidator() {
+        pattern = Pattern.compile(EMAIL_PATTERN);
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/controllers/EmailValidator.java]: This field is also can be static

File: server-app/src/main/java/ua/com/findcoach/domain/Coach.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package ua.com.findcoach.domain;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name="coach")
+@DiscriminatorColumn(name="is_coach", discriminatorType = DiscriminatorType.INTEGER)
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/domain/Coach.java]: https://wiki.eclipse.org/EclipseLink/UserGuide/JPA/Basic_JPA_Development/Entities/Inheritance#Example:_Using_a_DiscriminatorColumn_with_a_boolean_type_do_define_inheritance

Check this out and use discriminator column for boolean type

File: server-app/pom.xml [hunk 15]
```diff
@@ -9,14 +9,117 @@
...
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>3.1.0</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-core</artifactId>
+            <version>2.6.1</version>
```
Comment (senichas) [inline server-app/pom.xml]: Maven property

File: server-app/src/main/java/ua/com/findcoach/controllers/AuthenticationController.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    @ResponseBody
+    public AuthentificationResponse postAnswer(@RequestBody String body) throws JsonMappingException, JsonParseException, IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        AuthentificationResponse authentificationResponse = new AuthentificationResponse();
+        String decodeJSON = new URLDecoder().decode(body, "UTF-8");
+        AuthenticationRequest authenticationRequest = mapper.readValue(decodeJSON, AuthenticationRequest.class);
+        if (!emailValidator.validate(authenticationRequest.getEmail())) {
+            authentificationResponse.setMessage("You wrote wrong email");
+            authentificationResponse.setResult(false);
+            return authentificationResponse;
```
Comment (senichas) [inline server-app/src/main/java/ua/com/findcoach/controllers/AuthenticationController.java]: We can use one return statement. and assign needed values to response object in IF ELSE clause.

**Pull request merged**